components/properties/icu4x_shared.rs

```diff
@@ -0,0 +1,142 @@
+                $outer: Transparent<$inner>,
+            {
+                // Safety: Outer is repr(transparent) over Inner
+                // (enforced via trait safety invariant)
+                unsafe { &*(inner as *const $inner as *const $outer) }
+            }
+            /// Casts the inner type to the outer type.
+            ///
+            /// This function is safe, but it does not validate invariants
+            /// that the outer type might enforce. It is made available as
+            /// a private fn which can be called by another fn.
+            #[allow(dead_code)]
+            const fn cast_box_unchecked(
+                inner: alloc::boxed::Box<$inner>,
+            ) -> alloc::boxed::Box<$outer>
+            where
+                $outer: Transparent<$inner>,
+            {
+                // Safety: Outer is repr(transparent) over Inner
+                // (enforced via trait safety invariant)
+                unsafe { core::mem::transmute(inner) }
+            }
+        }
+    };
+}
+
+/// Implements `VarULE` on a `repr(transparent)` type.
+#[allow(unused_macros)]
+macro_rules! impl_transparent_varule {
+    ($outer:ident($inner:path)) => {
+        // Safety:
+        //
+        // 1. `repr(transparent)`, enforced by trait bound, implies no padding
+        // 2. `repr(transparent)`, enforced by trait bound, implies alignment 1
+        // 3. Composition of `repr(transparent)` `validate_by_slice` with
+        //    `validate_inner` from the `Transparent` trait implies
+        //    valid bytes
+        // 4. The `repr(transparent)` `validate_byte_slice` implies
+        //    that all bytes are covered
+        // 5. Composition of `repr(transparent)` `from_byte_slice_unchecked` with
+        //    `cast_ref_unchecked` retains the same reference
+        // 6. Other methods are left as default
+        // 7. Equality enforced via `Eq` bound
+        unsafe impl zerovec::ule::VarULE for $outer
+        where
+            $outer: Transparent<$inner> + Eq,
+        {
+            #[inline]
+            fn validate_byte_slice(bytes: &[u8]) -> Result<(), zerovec::ZeroVecError> {
+                let inner = <$inner>::parse_byte_slice(bytes)?;
+                Self::validate_inner(inner)
+                    .then_some(())
+                    .ok_or(zerovec::ZeroVecError::parse::<Self>())
+            }
+            #[inline]
+            unsafe fn from_byte_slice_unchecked(bytes: &[u8]) -> &Self {
+                let inner = <$inner>::from_byte_slice_unchecked(bytes);
+                Self::cast_ref_unchecked(inner)
+            }
+        }
+    };
+}
+
+/// Implements `serde::Deserialize` on a `repr(transparent)` type.
+#[allow(unused_macros)]
+macro_rules! impl_transparent_serde {
+    ($outer:ident($inner:path)) => {
+        impl serde::Serialize for $outer {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: serde::ser::Serializer,
```

Comment:
Note: This impl is not bound on `Transparent` because (1) it isn't needed for safety reasons, (2) Transparent is a private trait, and (3) the generated functions use `cast_ref_unchecked` which in effect enforces transparency since those functions require transparency